server/src/main/java/org/elasticsearch/index/mapper/MapperService.java

```diff
@@ -169,54 +170,55 @@ DocumentParser documentParser() {
-        } else {
-            logger.debug("[{}] {} mapping (source suppressed due to length, use TRACE level if needed)",
-                index(), op);
+        MappingMetadata newMappingMetadata = newIndexMetadata.mapping();
+        if (newMappingMetadata != null) {
+            String type = newMappingMetadata.type();
+            CompressedXContent incomingMappingSource = newMappingMetadata.source();
+            Mapping incomingMapping = parseMappings(type, incomingMappingSource);
+            DocumentMapper previousMapper;
+            synchronized (this) {
+                previousMapper = this.mapper;
+                assertRefreshIsNotNeeded(previousMapper, type, incomingMappingSource, incomingMapping);
+                this.mapper = newDocumentMapper(incomingMapping, MergeReason.MAPPING_RECOVERY);
+            }
+            String op = previousMapper != null ? "updated" : "added";
+            if (logger.isDebugEnabled() && incomingMappingSource.compressed().length < 512) {
+                logger.debug("[{}] {} mapping, source [{}]", index(), op, incomingMappingSource.string());
+            } else if (logger.isTraceEnabled()) {
+                logger.trace("[{}] {} mapping, source [{}]", index(), op, incomingMappingSource.string());
+            } else {
+                logger.debug("[{}] {} mapping (source suppressed due to length, use TRACE level if needed)",
+                    index(), op);
+            }
         }
+    }
 
-        // refresh mapping can happen when the parsing/merging of the mapping from the metadata doesn't result in the same
-        // mapping, in this case, we send to the master to refresh its own version of the mappings (to conform with the
-        // merge version of it, which it does when refreshing the mappings), and warn log it.
-        if (documentMapper().mappingSource().equals(incomingMappingSource) == false) {
-            logger.debug("[{}] parsed mapping, and got different sources\noriginal:\n{}\nparsed:\n{}",
-                index(), incomingMappingSource, documentMapper().mappingSource());
-
-            requireRefresh = true;
+    private void assertRefreshIsNotNeeded(DocumentMapper currentMapper,
+                                          String type,
+                                          CompressedXContent incomingMappingSource,
+                                          Mapping incomingMapping) {
+        if (Assertions.ENABLED) {
+            Mapping mergedMapping = mergeMappings(currentMapper, incomingMapping, MergeReason.MAPPING_RECOVERY);
+            CompressedXContent mergedMappingSource;
+            try {
+                mergedMappingSource = new CompressedXContent(mergedMapping, XContentType.JSON, ToXContent.EMPTY_PARAMS);
+            } catch (Exception e) {
+                throw new ElasticsearchGenerationException("failed to serialize source for type [" + type + "]", e);
```

Comment:
Is that acceptable or should we throw an assertion error here too?

Reply:
good question! it looks pretty bad, and we are only executing if assertions are enabled. Possibly throwing assertion makes sense to me